### ISSUE TITLE
"ReferenceError: h$doneMain is not defined" during booting

### ISSUE THREAD

**Issue by k-bx** [ghcjs/ghcjs#374]
Steps to reproduce:

- launch new Ubuntu 14.04 Vagrant-based image
- install ghc-7.10.2 from [Herbert's PPA](https://launchpad.net/~hvr/+archive/ubuntu/ghc), node 0.12.7
- clone & checkout ghcjs and ghcjs-prim into `improved-base` branches, install with `cabal install ./ghcjs ./ghcjs-prim`
- run `ghcjs-boot --dev --ghcjs-boot-dev-branch ghc-7.10`

Output I see (the end of it):

```
[72 of 79] Compiling Distribution.Simple.Build ( boot/cabal/Cabal/Distribution/Simple/Build.hs, boot/cabal/Cabal/dist/setup/Distribution/Simple/Build.js_o )
[73 of 79] Compiling Distribution.Simple.Install ( boot/cabal/Cabal/Distribution/Simple/Install.hs, boot/cabal/Cabal/dist/setup/Distribution/Simple/Install.js_o )
[74 of 79] Compiling Distribution.Simple.Haddock ( boot/cabal/Cabal/Distribution/Simple/Haddock.hs, boot/cabal/Cabal/dist/setup/Distribution/Simple/Haddock.js_o )
[75 of 79] Compiling Distribution.PackageDescription.Check ( boot/cabal/Cabal/Distribution/PackageDescription/Check.hs, boot/cabal/Cabal/dist/setup/Distribution/PackageDescription/Check.js_o )
[76 of 79] Compiling Distribution.Simple.SrcDist ( boot/cabal/Cabal/Distribution/Simple/SrcDist.hs, boot/cabal/Cabal/dist/setup/Distribution/Simple/SrcDist.js_o )
[77 of 79] Compiling Distribution.Simple.Configure ( boot/cabal/Cabal/Distribution/Simple/Configure.hs, boot/cabal/Cabal/dist/setup/Distribution/Simple/Configure.js_o )
[78 of 79] Compiling Distribution.Simple ( boot/cabal/Cabal/Distribution/Simple.hs, boot/cabal/Cabal/dist/setup/Distribution/Simple.js_o )
[79 of 79] Compiling Main             ( boot/cabal/Cabal/dist/setup/setup.hs, boot/cabal/Cabal/dist/setup/Main.js_o )
Linking ./boot/cabal/Cabal/dist/setup/setup.jsexe (Distribution.Compat.Binary,Distribution.Compat.CopyFile,Distribution.Compat.CreatePipe,Distribution.Compat.Environment,Distribution.Compat.Exception,Distribution.Compat.ReadP,Distribution.Compat.TempFile,Distribution.Compiler,Distribution.GetOpt,Distribution.InstalledPackageInfo,Distribution.License,Distribution.ModuleName,Distribution.Package,Distribution.PackageDescription,Distribution.PackageDescription.Check,Distribution.PackageDescription.Configuration,Distribution.PackageDescription.Parse,Distribution.PackageDescription.Utils,Distribution.ParseUtils,Distribution.ReadE,Distribution.Simple,Distribution.Simple.Bench,Distribution.Simple.Build,Distribution.Simple.Build.Macros,Distribution.Simple.Build.PathsModule,Distribution.Simple.BuildPaths,Distribution.Simple.BuildTarget,Distribution.Simple.CCompiler,Distribution.Simple.Command,Distribution.Simple.Compiler,Distribution.Simple.Configure,Distribution.Simple.GHC,Distribution.Simple.GHC.IPI641,Distribution.Simple.GHC.IPI642,Distribution.Simple.GHC.ImplInfo,Distribution.Simple.GHC.Internal,Distribution.Simple.GHCJS,Distribution.Simple.Haddock,Distribution.Simple.HaskellSuite,Distribution.Simple.Hpc,Distribution.Simple.Install,Distribution.Simple.InstallDirs,Distribution.Simple.JHC,Distribution.Simple.LHC,Distribution.Simple.LocalBuildInfo,Distribution.Simple.PackageIndex,Distribution.Simple.PreProcess,Distribution.Simple.PreProcess.Unlit,Distribution.Simple.Program,Distribution.Simple.Program.Ar,Distribution.Simple.Program.Builtin,Distribution.Simple.Program.Db,Distribution.Simple.Program.Find,Distribution.Simple.Program.GHC,Distribution.Simple.Program.HcPkg,Distribution.Simple.Program.Hpc,Distribution.Simple.Program.Ld,Distribution.Simple.Program.Run,Distribution.Simple.Program.Script,Distribution.Simple.Program.Strip,Distribution.Simple.Program.Types,Distribution.Simple.Register,Distribution.Simple.Setup,Distribution.Simple.SrcDist,Distribution.Simple.Test,Distribution.Simple.Test.ExeV10,Distribution.Simple.Test.LibV09,Distribution.Simple.Test.Log,Distribution.Simple.UHC,Distribution.Simple.UserHooks,Distribution.Simple.Utils,Distribution.System,Distribution.TestSuite,Distribution.Text,Distribution.Utils.NubList,Distribution.Verbosity,Distribution.Version,Language.Haskell.Extension,Main)
/home/vagrant/.ghcjs/x86_64-linux-0.1.0-7.10.2/ghcjs/ghcjs-boot/boot/cabal/Cabal/dist/setup/setup:13504
h$o(h$doneMain, (-1), 0, 0, 256, null);
    ^
ReferenceError: h$doneMain is not defined
    at Object.<anonymous> (/home/vagrant/.ghcjs/x86_64-linux-0.1.0-7.10.2/ghcjs/ghcjs-boot/boot/cabal/Cabal/dist/setup/setup:13504:5)
    at Module._compile (module.js:460:26)
    at Object.Module._extensions..js (module.js:478:10)
    at Module.load (module.js:355:32)
    at Function.Module._load (module.js:310:12)
    at Function.Module.runMain (module.js:501:10)
    at startup (node.js:129:16)
    at node.js:814:3
Failed to install Cabal-1.22.4.0
cabal: Error: some packages failed to install:
Cabal-1.22.4.0 failed during the configure step. The exception was:
ExitFailure 1
ghcjs-boot: error running: /opt/cabal/1.22/bin/cabal --config-file /home/vagrant/.ghcjs/x86_64-linux-0.1.0-7.10.2/ghcjs/cabalBootConfig --ignore-sandbox install ./boot/cabal/Cabal --global --ghcjs --one-shot --avoid-reinstalls --builddir dist --with-compiler /home/vagrant/.cabal/bin/ghcjs --with-hc-pkg /home/vagrant/.cabal/bin/ghcjs-pkg --prefix /home/vagrant/.ghcjs/x86_64-linux-0.1.0-7.10.2/ghcjs --enable-documentation --haddock-html --haddock-hyperlink-source --enable-shared --enable-library-profiling --root-cmd /home/vagrant/.ghcjs/x86_64-linux-0.1.0-7.10.2/ghcjs/run.sh
exit status: 1
stderr: Warning: No remote package servers have been specified. Usually you would have
one specified in the config file.
/home/vagrant/.ghcjs/x86_64-linux-0.1.0-7.10.2/ghcjs/ghcjs-boot/boot/cabal/Cabal/dist/setup/setup:13504
h$o(h$doneMain, (-1), 0, 0, 256, null);
    ^
ReferenceError: h$doneMain is not defined
    at Object.<anonymous> (/home/vagrant/.ghcjs/x86_64-linux-0.1.0-7.10.2/ghcjs/ghcjs-boot/boot/cabal/Cabal/dist/setup/setup:13504:5)
    at Module._compile (module.js:460:26)
    at Object.Module._extensions..js (module.js:478:10)
    at Module.load (module.js:355:32)
    at Function.Module._load (module.js:310:12)
    at Function.Module.runMain (module.js:501:10)
    at startup (node.js:129:16)
    at node.js:814:3
cabal: Error: some packages failed to install:
Cabal-1.22.4.0 failed during the configure step. The exception was:
ExitFailure 1
```

Please help!

**Comment by pavelkogan** [comment]
Same here, with node 0.10.38

**Comment by mgsloan** [comment]
This is also the issue I'm encountering with booting.  I'm using Ubuntu 14.10, with a fresh install of the following:

GHC 7.10.2
cabal-install-1.22.6.0
Cabal-1.22.4.0
HEAD of ghcjs improved base https://github.com/ghcjs/ghcjs/commit/7a718363ee1584e9ae43b49679b560099df79f86
HEAD of ghcjs-prim improved base https://github.com/ghcjs/ghcjs-prim/commit/02cd68b9a1000a27149d6a56eaff618e2f77c610

**Comment by kfigiela** [comment]
First, it seems that `ghcjs-prim` was moved from separate repo to `ghcjs` repo (https://github.com/ghcjs/ghcjs/tree/improved-base/lib/ghcjs-prim).
Second, do you ghcjs-boot with `improved-base` of `shims`?

I've just installed ghcjs on vanilla Ubuntu 14.10 VM with
* GHC 7.10.2
* cabal-install-1.22.6.0
* Cabal-1.22.4.0
* GHCJS from `improved-base` @ 7a718363ee1584e9ae43b49679b560099df79f86 
* ghcjs-boot & shims from `improved-base` branches (`ghcjs-boot --dev --no-prof --no-haddock --ghcjs-boot-dev-branch improved-base --shims-dev-branch improved-base`)

**Comment by mgsloan** [comment]
@kfigiela awesome, that did the trick!

It was a bit tricky, though, in case it is useful to others, here are the things I tried:

* Using @kfigiela 's ghcjs-boot invocation above.  This gave me the same error.

* Removed ~/.ghcjs and tried again.  Still the same error :/

* Using the ghcjs-boot invocation but with `--clean`.  This time I got an error while the boot script built `ghcjs-prim`.

* Now the way you started your comment is making sense to me.  Based on @kfigiela's instructions here, https://github.com/commercialhaskell/stack/issues/337#issuecomment-128672097 , I tried using the `lib/ghcjs-prim` within the `ghcjs`, instead of using `ghcjs`.  Success!!!

So, in summary:

* Follow @kfigiela 's instructions [here](https://github.com/commercialhaskell/stack/issues/337#issuecomment-128672097) and if booting previously failed, use `--clean`.

**Comment by bergey** [comment]
@kfigiela Thanks, that worked for me, too!

**Comment by kfigiela** [comment]
@mgsloan , @bergey: Good to hear that! 

@luite I took my freedom and updated wiki page (https://github.com/ghcjs/ghcjs/wiki/GHCJS-with-GHC-7.10) with my instructions

**Comment by k-bx** [comment]
I guess this is outdated and works now. Closing. Please reopen if necessary.
